### ISSUE TITLE
Trunk broken with copy_out_and_err

### ISSUE THREAD

**Issue by gaow** [vatlab/SoS#836]
I cannot easily provide a MWE for this one:

```
  File "/home/gaow/Public/miniconda3/lib/python3.6/site-packages/sos-0.9.9.9-py3.6.egg/sos/sos_task.py", line 341, in _execute_task
    copy_out_and_err(res)
Traceback (most recent call last):
  File "/home/gaow/Public/miniconda3/lib/python3.6/site-packages/sos-0.9.9.9-py3.6.egg/sos/sos_task.py", line 341, in _execute_task
    copy_out_and_err(res)
  File "/home/gaow/Public/miniconda3/lib/python3.6/site-packages/sos-0.9.9.9-py3.6.egg/sos/sos_task.py", line 307, in copy_out_and_err
    out.write('output: {}\n'.format(result['output']).encode())
KeyError: 'output'
ERROR: M1_b09355e993a6cbfa99176fb5d39e741c ``failed`: 'output'
```

Not sure if it is obvious before I work harder for an MWE?

**Comment by BoPeng** [comment]
Not sure if I am looking at the correct place but all task results are supposed to have `output` (https://github.com/vatlab/SoS/blob/master/src/sos/sos_task.py#L258). Let us dig a bit more.

**Comment by BoPeng** [comment]
I see. The problem should be at https://github.com/vatlab/SoS/blob/master/src/sos/sos_task.py#L574, that is to say, when one of the subtasks raises an exception, the result does not have `output`.

**Comment by gaow** [comment]
I think that's the case! I did not even realize my subtask may have exceptions thus cannot provide MWE.

**Comment by BoPeng** [comment]
Try it and let me know what you see.

**Comment by gaow** [comment]
I see the same:

```
ERROR: M1_543d19dec8928887e34c9fae6db17c54 ``failed`: Use "sos status 543d19dec8928887e34c9fae6db17c54 -v4" for details.
ERROR: 'output'
```

I tried to make MWE by putting together tasks doomed to fail, yet still cannot reproduce it.

**Comment by BoPeng** [comment]
Can you just print what that `result` contains? We are running into the ugly part of Python where anything can be passed to anywhere without type check.

**Comment by gaow** [comment]
I think your earlier post pointed out correctly what the problem is (exceptions occurred) but the patch did not fix it. Here is output of `print(results)`

```
[{'ret_code': 1, 'exception': RuntimeError('Use "sos status 543d19dec8928887e34c9fae6db17c54 -v4" for details.',)}]
```

**Comment by BoPeng** [comment]
Just pushed some more but I have to work on something else now. If you can `print`, you should be able to see why the `exception` is not raised by my patch. Could you check it?

**Comment by gaow** [comment]
This patch seems to make sense and has removed the `KeyError` -- looks great! I'll yet have to fix up my task and see how it behaves when tasks do not fail.

**Comment by gaow** [comment]
The error I know get is: 

```
FileNotFoundError: [Errno 2] No such file or directory: '/home/gaow/.sos/tasks/127e65942ac02d926a747959f03986c9.task'
```

after I fix my task ...

**Comment by BoPeng** [comment]
backtrace?

**Comment by gaow** [comment]
Updated. working on MWE now.

**Comment by BoPeng** [comment]
OK, just from the surrounding code, there is a possibility that the task file does not exist. Not sure why, but I have moved the new task duration code after checking the existence of task file.

**Comment by gaow** [comment]
It's worse: still fails, but much less error message

```
$ sos status 6d90b894b0294810e7b7607e3ce15aba -v4
TRACE: /opt/miniconda3/lib/python3.6/site-packages/sos-0.9.9.9-py3.6.egg/sos/site_config.yml does not exist
6d90b894b0294810e7b7607e3ce15aba	non-exist

```

**Comment by BoPeng** [comment]
So do you have this `~/.sos/tasks/6d90b894b0294810e7b7607e3ce15aba.task` file?

**Comment by gaow** [comment]
> So do you have this ~/.sos/tasks/6d90b894b0294810e7b7607e3ce15aba.task file?

No I do not.

I think the issue is still that `sos status` does not provide informative information. If I knowingly break my code, this is what I get with a test example:

```
INFO: Executing default_0: 
INFO: input:    []
ERROR: 6049500e735ca4fbe845050f478e9d54 failed: Use "sos status 6049500e735ca4fbe845050f478e9d54 -v4" for details.
ERROR: M1_6049500e735ca4fbe845050f478e9d54 ``failed`: Use "sos status 6049500e735ca4fbe845050f478e9d54 -v4" for details.
INFO: M1_6049500e735ca4fbe845050f478e9d54.out:
6049500e735ca4fbe845050f478e9d54: failed
6049500e735ca4fbe845050f478e9d54: failed
output: []

INFO: M1_6049500e735ca4fbe845050f478e9d54.err:
6049500e735ca4fbe845050f478e9d54: failed
2017-10-26 20:13:14,762: DEBUG: 6049500e735ca4fbe845050f478e9d54 ``started``
2017-10-26 20:13:14,964: DEBUG: R library rhdf5 (2.20.0) is available
2017-10-26 20:13:15,059: DEBUG: R library tools (3.4.1) is available
2017-10-26 20:13:16,204: DEBUG: R library glmnet (2.0.10) is available
Error in h5read(genotype_file, geno_table) : 
  Object /ENSG00000162739 does not exist in this HDF5 file.
Calls: analyze -> load_data -> h5read
Execution halted
2017-10-26 20:13:16,548: DEBUG: Failed to execute R (interpolate(r"""
...
...

```
Then I know exactly what is broken. But I dislike it because there is too much information when my script is really long (not shown here). However this is not displayed until the end of my jobs. What in fact happens to me is that I send ~1000 jobs, I see these errors from the first few jobs, I then ctrl-C to try to debug via `sos status ... -v4` which is not very informative.

**Comment by BoPeng** [comment]
Still do not quite get it. What exactly caused the problem? `sos status` shows `non-exist`, which says task file does not exist, not created, pretty easy right?

**Comment by gaow** [comment]
> What exactly caused the problem? 

I suspect it is still because there are errors in my tasks. For a MWE, the error message you see above is displayed at the end of `sos run`, and because it is MWE I can see it properly displayed -- which I actually dislike because it's too much information. Ideally I only want a prompt on the screen saying something failed, and I can then use `sos status` to check why it fails. But you see when I do `sos status` on the failed task in the previous post, instead of showing me those information, it simply says "non-exists". 

[test.ipynb.zip](https://github.com/vatlab/SoS/files/1420557/test.ipynb.zip)

Here is an example. You should be able to run it (maybe you can disable `depends` for now) and see what I meant, and what ideally one would expect.

**Comment by gaow** [comment]
Was this fixed? Somehow when I tried to use `sos status` there are now more error messages ...

**Comment by BoPeng** [comment]
Not yet... fixing stuff at home now. Will try it tonight.

**Comment by gaow** [comment]
Hmm interesting. So maybe for particular errors `sos status` will show messages? Let me keep trying looking for runs with error message that does not show up in `sos status` (apparently there are multiple types of corner cases in my procedure when I apply it in parallel to ~50K tasks)

**Comment by BoPeng** [comment]
There is a strange problem that the final error message from `logger` will clear the `.err` file, so in the end we only see the useless `use sos status -v4` message. Will investigate more.

**Comment by BoPeng** [comment]
Are you still seeing more error messages?

**Comment by gaow** [comment]
With the latest trunk, for that MWE I uploaded, I see what I saw before: overly comprehensive (too much!) error at the end of the run, yet non-informative `sos status` result.

**Comment by BoPeng** [comment]
I do see something like

```
Error in h5checktypeOrOpenLoc(file, readonly = TRUE) :
  Error in h5checktypeOrOpenLoc(). Cannot open file. File '/home/gaow/Documents/GTExV8/GTExV8.genotype.cis.bad.h5' does not exist.
Calls: analyze -> load_data -> h5read -> h5checktypeOrOpenLoc
Execution halted
``` 
on my end from the `sos status -v4` output, although it is buried in a long `.err` file.

**Comment by gaow** [comment]
Really? then this is weird, because I see:

```
$ sos status b3b6a1ff29d09e7e1294fdd47f03e83b -v4
TRACE: /home/gaow/Public/miniconda3/lib/python3.6/site-packages/sos-0.9.9.9-py3.6.egg/sos/site_config.yml does not exist
b3b6a1ff29d09e7e1294fdd47f03e83b	non-exist
```

The long `err` file gets printed at the end of execution, on the screen, that buries the info you identified. That's another issue -- better only keep the essential error.

**Comment by BoPeng** [comment]
I am running `sos status` on the master task `M1_d290...`. Are you trying to get information on a subtask?

**Comment by gaow** [comment]
Ohh I see the difference now. I was just copy-paste the error prompt:

```
2017-10-30 21:11:55,280: ERROR: b3b6a1ff29d09e7e1294fdd47f03e83b ``failed``: Use "sos status b3b6a1ff29d09e7e1294fdd47f03e83b -v4" for details.
```

which might be a bit misleading, and not informative ...

**Comment by BoPeng** [comment]
The error messages for subtasks are actually aggregated to the master task... I will see what I can do if you are indeed trying to get the error message of subtasks.

**Comment by gaow** [comment]
I guess it does not matter where I get the message, but rather be given the proper direction and be able to easily pin-point to the relevant error message one way or another. So any design that makes this easier would be good in practice.

**Comment by BoPeng** [comment]
At least I now see the problem, let me have a look.

**Comment by gaow** [comment]
I'm testing #752 now but is this going to be improved for the new release?

**Comment by BoPeng** [comment]
I do not yet know how to improve on this.

**Comment by gaow** [comment]
I think it is good enough if we:

1. Change the error message: `Use "sos status <sub_task_id> ..." ...` to `sos status <master_task_id> ..." ...` so that the proper message can be displayed
2. Do not display the entire script of the action in question, only show the error message captured
3. Do not display concrete error message at the end of execution (because it will be messy for thousands of jobs), but only display the `sos status` tip.

**Comment by BoPeng** [comment]
Just as an example, error messages of 

```
[10]
input: for_each={'i': range(10)}
task: trunk_size=5
sh:
  eho  ${i}
```

are

```
$ sos run test -s force
INFO: Executing default_10:
INFO: input:    []
ERROR: fdc03ab5a7ef79537fa942b800a02820 failed: Use "sos status fdc03ab5a7ef79537fa942b800a02820 -v4" for details.
ERROR: M5_fdc03ab5a7ef79537fa942b800a02820 ``failed`: Use "sos status fdc03ab5a7ef79537fa942b800a02820 -v4" for details.
ERROR: 987a348674f5d87bcb3222122bd3bb65 failed: Use "sos status 987a348674f5d87bcb3222122bd3bb65 -v4" for details.
ERROR: M5_987a348674f5d87bcb3222122bd3bb65 ``failed`: Use "sos status 987a348674f5d87bcb3222122bd3bb65 -v4" for details.
INFO: M5_fdc03ab5a7ef79537fa942b800a02820.out:
fdc03ab5a7ef79537fa942b800a02820: failed
fdc03ab5a7ef79537fa942b800a02820: failed
output: []

INFO: M5_fdc03ab5a7ef79537fa942b800a02820.err:
fdc03ab5a7ef79537fa942b800a02820: failed
2017-10-31 22:22:55,573: DEBUG: fdc03ab5a7ef79537fa942b800a02820 ``started``
/var/folders/ys/gnzk0qbx5wbdgm531v82xxljv5yqy8/T/tmp2kwla8d2.sh: line 1: eho: command not found
2017-10-31 22:22:55,642: DEBUG: Failed to execute sh (interpolate(r"""eho  ${i}
""", '${ }', locals()))
2017-10-31 22:22:55,643: ERROR: fdc03ab5a7ef79537fa942b800a02820 ``failed``: Use "sos status fdc03ab5a7ef79537fa942b800a02820 -v4" for details.
Task M5_fdc03ab5a7ef79537fa942b800a02820 exits with code 1
INFO: M5_987a348674f5d87bcb3222122bd3bb65.out:
987a348674f5d87bcb3222122bd3bb65: failed
987a348674f5d87bcb3222122bd3bb65: failed
output: []

INFO: M5_987a348674f5d87bcb3222122bd3bb65.err:
987a348674f5d87bcb3222122bd3bb65: failed
2017-10-31 22:22:55,655: DEBUG: 987a348674f5d87bcb3222122bd3bb65 ``started``
/var/folders/ys/gnzk0qbx5wbdgm531v82xxljv5yqy8/T/tmptcksbppz.sh: line 1: eho: command not found
2017-10-31 22:22:55,696: DEBUG: Failed to execute sh (interpolate(r"""eho  ${i}
""", '${ }', locals()))
2017-10-31 22:22:55,696: ERROR: 987a348674f5d87bcb3222122bd3bb65 ``failed``: Use "sos status 987a348674f5d87bcb3222122bd3bb65 -v4" for details.
Task M5_987a348674f5d87bcb3222122bd3bb65 exits with code 1
ERROR: 2 jobs failed
```

**Comment by gaow** [comment]
Right, this line:

```
Failed to execute sh (interpolate(r"""eho  ${i}
""", '${ }', locals()))
```

could be annoyingly long if the action contains a lot of codes (eg large chunk of Python / R).

**Comment by BoPeng** [comment]
Not sure how much better it is but I have
1. removed duplicated error message from subtask and master task
2. remove "please use sos status etc" because the action cannot tell if it is from a subtask.
3. use `short_repr` for script (`Failed to execute`).

```
$ sos run test -s force
INFO: Executing default_10:
INFO: input:    []
INFO: M5_fdc03ab5a7ef79537fa942b800a02820 restart from status failed
INFO: M5_987a348674f5d87bcb3222122bd3bb65 restart from status failed
ERROR: fdc03ab5a7ef79537fa942b800a02820 failed: Failed to execute commmand /bin/sh /Users/bpeng1/sos/.sos/default_10_0_a57de2c5.sh (ret=127, workdir=/Users/bpeng1/sos, task=fdc03ab5a7ef79537fa942b800a02820)
ERROR: M5_fdc03ab5a7ef79537fa942b800a02820 failed due to failure of subtask fdc03ab5a7ef79537fa942b800a02820
ERROR: 987a348674f5d87bcb3222122bd3bb65 failed: Failed to execute commmand /bin/sh /Users/bpeng1/sos/.sos/default_10_5_87f837c7.sh (ret=127, workdir=/Users/bpeng1/sos, task=987a348674f5d87bcb3222122bd3bb65)
ERROR: M5_987a348674f5d87bcb3222122bd3bb65 failed due to failure of subtask 987a348674f5d87bcb3222122bd3bb65
INFO: M5_fdc03ab5a7ef79537fa942b800a02820.out:
fdc03ab5a7ef79537fa942b800a02820: failed
fdc03ab5a7ef79537fa942b800a02820: failed
output: []

INFO: M5_fdc03ab5a7ef79537fa942b800a02820.err:
fdc03ab5a7ef79537fa942b800a02820: failed
2017-10-31 23:14:42,214: DEBUG: fdc03ab5a7ef79537fa942b800a02820 ``started``
/var/folders/ys/gnzk0qbx5wbdgm531v82xxljv5yqy8/T/tmp57kp0c57.sh: line 1: eho: command not found
2017-10-31 23:14:42,253: DEBUG: Failed to execute 'sh (interpolate(r"""eho  ${i}\n""", \'${ }\', locals()))'
2017-10-31 23:14:42,253: ERROR: fdc03ab5a7ef79537fa942b800a02820 ``failed``: Failed to execute commmand ``/bin/sh /Users/bpeng1/sos/.sos/default_10_0_a57de2c5.sh`` (ret=127, workdir=/Users/bpeng1/sos, task=fdc03ab5a7ef79537fa942b800a02820)
Task M5_fdc03ab5a7ef79537fa942b800a02820 exits with code 1
INFO: M5_987a348674f5d87bcb3222122bd3bb65.out:
987a348674f5d87bcb3222122bd3bb65: failed
987a348674f5d87bcb3222122bd3bb65: failed
output: []

INFO: M5_987a348674f5d87bcb3222122bd3bb65.err:
987a348674f5d87bcb3222122bd3bb65: failed
2017-10-31 23:14:42,312: DEBUG: 987a348674f5d87bcb3222122bd3bb65 ``started``
/var/folders/ys/gnzk0qbx5wbdgm531v82xxljv5yqy8/T/tmpail6t791.sh: line 1: eho: command not found
2017-10-31 23:14:42,355: DEBUG: Failed to execute 'sh (interpolate(r"""eho  ${i}\n""", \'${ }\', locals()))'
2017-10-31 23:14:42,355: ERROR: 987a348674f5d87bcb3222122bd3bb65 ``failed``: Failed to execute commmand ``/bin/sh /Users/bpeng1/sos/.sos/default_10_5_87f837c7.sh`` (ret=127, workdir=/Users/bpeng1/sos, task=987a348674f5d87bcb3222122bd3bb65)
Task M5_987a348674f5d87bcb3222122bd3bb65 exits with code 1
ERROR: 2 jobs failed
```

**Comment by gaow** [comment]
It's a bit improvement.  I guess ideally all lines with `2017-10-31` should not be shown on the screen though should display in the `err` file. Is that very difficult to achieve? That way the error prompts will be much neater when many tasks present.

**Comment by BoPeng** [comment]
The `2017-10-31` lines are from the `.err` file, which the error message tries to display so that you do not have to use commands such as `sos status` to show it. I can replace them with something like `Complete log file is available at .....err`.

**Comment by gaow** [comment]
Exactly. The part of the actual error is relevant, but it is also buried in the err file ... Is there a way to only remove those Python `logger` generated lines when displaying summary of `err` file?
 If not, then compared to seeing messy screen output I would rather go check the `err` file manually. Of course I may be a bit biased because I often deal with many tasks; only neat screen output makes things appear more managable ...

**Comment by BoPeng** [comment]
Something like the following?

```
INFO: Executing default_10:
INFO: input:    []
INFO: M5_fdc03ab5a7ef79537fa942b800a02820 restart from status failed
INFO: M5_987a348674f5d87bcb3222122bd3bb65 restart from status failed
ERROR: fdc03ab5a7ef79537fa942b800a02820 failed: Failed to execute commmand /bin/sh /Users/bpeng1/sos/.sos/default_10_0_66904347.sh (ret=127, workdir=/Users/bpeng1/sos, task=fdc03ab5a7ef79537fa942b800a02820)
ERROR: M5_fdc03ab5a7ef79537fa942b800a02820 failed due to failure of subtask fdc03ab5a7ef79537fa942b800a02820
ERROR: 987a348674f5d87bcb3222122bd3bb65 failed: Failed to execute commmand /bin/sh /Users/bpeng1/sos/.sos/default_10_5_486da9fd.sh (ret=127, workdir=/Users/bpeng1/sos, task=987a348674f5d87bcb3222122bd3bb65)
ERROR: M5_987a348674f5d87bcb3222122bd3bb65 failed due to failure of subtask 987a348674f5d87bcb3222122bd3bb65

~/.sos/tasks/M5_fdc03ab5a7ef79537fa942b800a02820.out:
fdc03ab5a7ef79537fa942b800a02820: failed
fdc03ab5a7ef79537fa942b800a02820: failed
output: []

~/.sos/tasks/M5_fdc03ab5a7ef79537fa942b800a02820.err:
fdc03ab5a7ef79537fa942b800a02820: failed
/var/folders/ys/gnzk0qbx5wbdgm531v82xxljv5yqy8/T/tmp8o82zjfj.sh: line 1: eho: command not found
Task M5_fdc03ab5a7ef79537fa942b800a02820 exits with code 1

~/.sos/tasks/M5_987a348674f5d87bcb3222122bd3bb65.out:
987a348674f5d87bcb3222122bd3bb65: failed
987a348674f5d87bcb3222122bd3bb65: failed
output: []

~/.sos/tasks/M5_987a348674f5d87bcb3222122bd3bb65.err:
987a348674f5d87bcb3222122bd3bb65: failed
/var/folders/ys/gnzk0qbx5wbdgm531v82xxljv5yqy8/T/tmpjl52opl5.sh: line 1: eho: command not found
Task M5_987a348674f5d87bcb3222122bd3bb65 exits with code 1
ERROR: 2 jobs failed
```

**Comment by gaow** [comment]
This is great!
